flink-runtime/src/main/java/org/apache/flink/runtime/jobgraph/jsonplan/JsonPlanGenerator.java

```diff
@@ -117,10 +120,25 @@ public static String generatePlan(
                 // write the core properties
                 JobVertexID vertexID = vertex.getID();
                 int storeParallelism = vertexParallelism.getParallelism(vertexID);
+                int parallelism =
+                        storeParallelism != -1 ? storeParallelism : vertex.getParallelism();
                 gen.writeStringField("id", vertexID.toString());
-                gen.writeNumberField(
-                        "parallelism",
-                        storeParallelism != -1 ? storeParallelism : vertex.getParallelism());
+                gen.writeNumberField("parallelism", parallelism);
+                int storeMaxParallelism =
+                        vertexParallelism
+                                .getMaxParallelismForVertices()
+                                .getOrDefault(vertexID, MAX_PARALLELISM_DEFAULT);
+                int maxParallelism =
+                        storeMaxParallelism == MAX_PARALLELISM_DEFAULT
+                                ? (vertex.getMaxParallelism() == MAX_PARALLELISM_DEFAULT
+                                        ? (parallelism == MAX_PARALLELISM_DEFAULT
```

Comment:
The condition should be `parallelism == ExecutionConfig#PARALLELISM_DEFAULT`. The same value `-1` can have different meanings.